stellargraph/core/utils.py

```diff
@@ -32,18 +32,22 @@ def is_real_iterable(x):
     return isinstance(x, collections.abc.Iterable) and not isinstance(x, (str, bytes))
 
 
-def normalize_adj(adj, symmetric=True):
+def normalize_adj(adj, symmetric=True, add_self_loops=False):
     """
     Normalize adjacency matrix.
 
     Args:
         adj: adjacency matrix
         symmetric: True if symmetric normalization or False if left-only normalization
-
+        add_self_loops: True if self loops are to be added before normalization, i.e., use A+I where A is the adjacency
+            matrix and I is a square identity matrix of the same size as A.
     Returns:
         Return a sparse normalized adjacency matrix.
     """
 
+    if add_self_loops:
+        adj = adj + sp.diags(np.ones(adj.shape[0]) - adj.diagonal())
```

Comment:
What's the reason for subtracting off existing self-loops? The papers I can remember all just use `A + I`, not `A + I - diag(A)` or similar, and e.g. the GCN code uses just `A + I` too (https://github.com/tkipf/gcn/blob/dab6e6b0d936c6590fb1a1b3f8fd7cef9fc4c858/gcn/utils.py#L134). The doc-string is currently slightly misleading too.

```suggestion
        adj = adj + sp.diags(np.ones(adj.shape[0]))
```

(This is a question so I can understand the approach, as I see that it's duplicated in several other places in this file too.)

Reply:
In all the papers, it is assumed (as far as I understand) that the input graph has no self loops. So, the operation A+I correctly adds a self loop for every node in the graph. The reason for this operation is to include the node's features in the calculation. For most datasets we use, e.g., Cora, MUTAG, PROTEINS, it is true that all nodes do not have self loops and hence, A+I works correctly. The problem arises if someone has a dataset with some nodes already having self loops or even all nodes already having self loops; this could be the case for custom datasets. Then A+I increases the weights for the nodes' self features unnecessarily. I don't know if having a 2 instead of 1 will really make a big difference but being careful about it seems reasonable. I suspect papers use A+I in all the equations for convenience; the practical implementation of this operation needs to be done a bit more carefully.

Our own GCN normalisation routine `core.utils.GCN_Aadj_feats_op` (which I have to admit I don't know if we still use) uses the same calculation and this is where the above line comes from in the first place.

Reply:
Fair enough.

> The reason for this operation is to include the node's features in the calculation

A different framing is that `A + I` is emphasising the self-nodes features rather than just including, in which case doing `A + I` directly seems like a reasonable interpretation. (I think [GIN](https://arxiv.org/abs/1810.00826) effectively does this, via `A + (1 + eps) * I` for some trainable (or not) `eps`.)

What do you think?

> the same calculation and this is where the above line comes from in the first place

Indeed: a better way to do this would've been to move (rather than copy) that calculation, and have `GCN_Aadj_feats_op` call `normalize_adj(..., add_self_loops=True)`, but this can be done elsewhere/elsewhen.

> which I have to admit I don't know if we still use

For future reference, this sort of question can be easily answered (in the affirmative, in this case) with `git grep`:

```
$ git grep GCN_Aadj_feats_op
...
stellargraph/mapper/full_batch_generators.py:from ..core.utils import GCN_Aadj_feats_op, PPNP_Aadj_feats_op
stellargraph/mapper/full_batch_generators.py:            self.features, self.Aadj = GCN_Aadj_feats_op(
...
```

Reply:
> Fair enough.
> 
> > The reason for this operation is to include the node's features in the calculation
> 
> A different framing is that `A + I` is emphasising the self-nodes features rather than just including, in which case doing `A + I` directly seems like a reasonable interpretation. (I think [GIN](https://arxiv.org/abs/1810.00826) effectively does this, via `A + (1 + eps) * I` for some trainable (or not) `eps`.)
> 
> What do you think?
> 

This might be another reasonable way to explain the A+I operation. My above quoted statement is based on how it is described in prior-art such as for example the DGCNN paper states, "Our proposed graph convolution adopts a propagation D^-1A (with A=A+I, my comment) which is very similar to that of PK, except for preserving nodes' self-information over iterations."

Also Kipf's GCN paper states, "Here, A = A + I is the adjacency matrix of the undirected graph G with added self-connections." which does not imply this operation is meant to emphasise nodes' self features but just to include them. 

That said, GIN does indeed extend the concept of simply adding self-connections so...basically, I have no idea what the correct justification for A+I  should be!

Reply:
Yeah, that makes sense. Maybe we can (eventually) offer all the options and it's up to a user to decide which one applies to their graph.